Refs #34043 -- Added GitHub action to capture screenshots in Selenium tests.

sarahboyce opened the pull request:
This is a proposal to help us confidently make UI changes. :muscle: 
Testing UI is always hard and many of us would not describe ourselves as frontend developers. Adding a layer of visual testing to our automated test suite will not only make our lives easier but can open up possibilities to us such as tests to cover visual regressions.

In [ticket 34043](https://code.djangoproject.com/ticket/34043), there is a suggestion that we should have a test suite/tooling to keep track of UI changes automatically. 

This PR adds the screenshot tests to our GitHub actions. This is currently triggered when a PR has the `selenium` label.
You can see the screenshots (in a zip file at the bottom called screenshots) here: https://github.com/django/django/actions/runs/5386413722?pr=16963

The action "fails" because I do not have write permissions to the repo. please see this comment https://github.com/django/django/pull/16963#discussion_r1243194672. I have tested this against my fork where I do have write permissions. This should be resolved on merge.

smithdc1 commented:
I'm not sure if this is helpful? Maybe you've already seen it!

https://simonwillison.net/2022/Mar/10/shot-scraper/

sarahboyce commented:
> I'm not sure if this is helpful? Maybe you've already seen it!
> 
> https://simonwillison.net/2022/Mar/10/shot-scraper/

I haven't but it's quite similar (I am also using playwright here but using the python version because Python = :heart:), will take a look!
I am trying also take inspiration from Adam's blog (https://adamj.eu/tech/2023/03/17/django-parameterized-tests-model-admin-classes/) around testing the admin and what I want in the end is something that will give you just changed screenshots with before and after so hopefully the css changes are easier when seeing unexpected changes (visual regressions are just difficult)

As an example, you can download and see screenshots here in the Artifacts section at the bottom: https://github.com/django/django/actions/runs/5225589580, it's not quite what I want yet but you get the idea. Need to check with people if we think this adds value.

sarahboyce commented on .github/workflows/selenium_tests.yml:
```diff
@@ -0,0 +1,34 @@
+
+      - name: Run selenium tests for current PR
+        working-directory: ./tests/
+        run: python runtests.py --selenium=chrome --headless --parallel 2 admin_changelist admin_inlines admin_widgets forms_tests
```
This is stating which tests to run because without it I was getting database threading errors
```
django.db.utils.DatabaseError: DatabaseWrapper objects created in a thread can only be used in that same thread. The object with alias 'default' was created in thread id 139881739424192 and this is thread id 139880491378240.
```

I had no idea where this was coming from but this [blog article](https://www.dimagi.com/blog/debugging-a-thread-related-test-failure/) implied that something is being imported in the test discovery and causing this issue. By only stating the test folders which have selenium tests, this went away but it is not ideal as it's a people are unlikely to know they must update this when adding selenium tests to new test folders.
I can try to find a better solution but if anyone knows how to fix this that would be lovely.

smithdc1 commented:
I expect that there is value to move the selenium tests to GHA irrespective of the new screenshotting feature. 

Maybe it's worth splitting that change out from here to get that merged first? 

We can then review that the new action reflects what's in Jenkins before possibly adding a selenium job here in addition to the two on Jenkins.

@felixxm -- is my understanding that it is desired to move these tests from Jenkins correct? 🤔

ngnpope reviewed (commented):
Thank for looking into this @sarahboyce. I was actually thinking about this very recently in relation to a change I was thinking of making in the admin and how we'd be able to tell that it didn't break anything. So this is most timely 😄 

I see you've shown some examples of the deltas using `pixelmatch`. It'd also be nice if we can come up with a way to easily review the changes, whether that's posting a comment with the diff (although that might get very noisy) or adding an HTML page that can be published to GitHub pages with a link added via a comment to enable easy review?

In addition, how confident are we that all styling stuff is captured in the selenium tests that exist? Do we want an example project somewhere that exercises everything which we can use to run this visual testing against?

sarahboyce commented on django/contrib/admin/tests.py:
```diff
@@ -231,3 +232,18 @@ def has_css_class(self, selector, klass):
+        screenshot_name = os.path.join(
+            screenshot_directory, f"{self._testMethodName}.png"
+        )
+        self.selenium.set_window_size(1366, 784)
```
The default screen size of selenium was strange, so expanded it to a "standard" desktop size. This actually [might not be the most common](https://www.hobo-web.co.uk/best-screen-size/) one. Need to pick one and put an appropriate comment :+1: 
We can take multiple screenshots at different sizes. I would be tempted to either start with desktop or desktop and mobile rather than do too many too early. :thinking:

sarahboyce replied:
Another idea, similar to how `runtests.py` has the cli argument `--selenium chrome,firefox` (list of browsers) maybe I should do something similar like `--screenshots 1366x784,1920x1200` (list of resolutions) and then you can run this locally and check different screen-sizes and the ci just does 1 or 2. :thinking:

sarahboyce commented on django/contrib/admin/tests.py:
```diff
@@ -231,3 +232,18 @@ def has_css_class(self, selector, klass):
+            screenshot_directory, f"{self._testMethodName}.png"
+        )
+        self.selenium.set_window_size(1366, 784)
+        self.selenium.save_screenshot(screenshot_name)
```
Thank you, pushed an update! That's much nicer

sarahboyce commented:
As the end of a test might not give us a screenshot at the right point, made the screenshot taking explicitly part of the test.
@ngnpope also added a mobile screenshot :+1: 

I added screenshots for one of the test files to give us an idea. You can see the screenshots here: https://github.com/django/django/actions/runs/5280983768?pr=16963
How do we feel about this approach? Am I getting warmer or colder do we think? (@smithdc1)

felixxm commented:
I moved the last commit to the separate PR, see #16989.

sabderemane commented on django/contrib/admin/tests.py:
```diff
@@ -231,3 +233,24 @@ def has_css_class(self, selector, klass):
+            {"name": "mobile", "width": 360, "height": 800},
+        ]
+        for spec in screenshot_specs:
+            self.selenium.set_window_size(spec["width"], spec["height"])
```
If I'm not mistaken, we don't have screenshot with rtl direction in HTML. I think it will be interesting if we can have that.
I'm not completely familiar with selenium, do you think it's possible at this stage ?

sarahboyce replied:
It should be possible for sure and it's a good idea, do you think it's similar to mobile where we would want it for every page?

sabderemane replied:
I think so, but if it's too many screenshots to have/check on, maybe we could have screenshots on list page and detail page, IMO these are the most significant pages for rtl

knyghty replied:
IMO it's good to have everywhere because while I usually remember to take screens of RTL, I always forget to test it when making further commits.

sarahboyce replied:
Ok I added logic that you can take the screenshots also in RTL. You can view screenshots here: https://github.com/django/django/actions/runs/5359753461/
It's an attachment named screenshots at the bottom :+1:

sabderemane replied:
LGTM 🚀

sarahboyce commented on django/test/selenium.py:
```diff
@@ -133,3 +137,86 @@ def disable_implicit_wait(self):
+    screenshots = False
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
```
@adamchainz I took a lot of inspiration from your package [unittest-parametrize](https://github.com/adamchainz/unittest-parametrize) here :heart: 
Feel free to add your thoughts on how best to structure this :+1:

sarahboyce replied:
Also @ngnpope if you want to look at the PR again, I approach has changed quite a bit to accommodate different requirements and would be good to get some feedback :star:

adamchainz replied:
Glad to inspire!

adamchainz commented on django/test/selenium.py:
```diff
@@ -133,3 +137,86 @@ def disable_implicit_wait(self):
+        for name, func in list(cls.__dict__.items()):
+            if not cls.screenshots:
+                continue
+            if not isinstance(func, FunctionType):
+                continue
```
I suggest deleting this since the attribute can come only from the decorator
```suggestion
```

I know you copied it from unittest-parametrize but that needs to be a bit more defensive because of the chance of interaction with other code. Whilst I'd say don't add code here that's not necessary.

adamchainz commented on django/test/selenium.py:
```diff
@@ -133,3 +137,86 @@ def disable_implicit_wait(self):
+
+                setattr(cls, test.__name__, test)
+
+    def desktop(self):
+        self.window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(1280, 720)
+
+    def disable_desktop(self):
+        self.selenium.set_window_size(
+            self.window_size["width"], self.window_size["height"]
+        )
```
turn these into context managers? Otherwise failures can pollute other tests.

sarahboyce replied:
Yes! I was doing this going, why is this so ugly?! :joy: thank you :+1:

sarahboyce commented on tests/runtests.py:
```diff
@@ -747,6 +754,9 @@ def paired_tests(paired_test, options, test_labels, start_at, start_after):
         SeleniumTestCaseBase.headless = options.headless
         SeleniumTestCaseBase.browsers = options.selenium
+        if options.screenshots:
+            options.tags = ["screenshot"]
```
Not sure what you mean here, when I have `--screenshots` it only makes sense to run the tests with screenshots so I don't think it's worth running other selenium tests without screenshots for example.
This quite specific to how we run our tests, do you think I need to handle some case I'm missing?

adamchainz replied:
I mean if `--tags` has been specified along with `--screenshots`, there should be an error, similar to the above for checking that `--selenium` has been passed.

sarahboyce commented on .github/workflows/screenshots.yml:
```diff
@@ -0,0 +1,69 @@
+          comment-author: 'github-actions[bot]'
+          body-includes: Screenshots created
+
+      - name: Create or update comment
```
The GitHub action fails at this point because this PR comes from a fork and does not have write permissions for the repository. This should be resolved if the PR is accepted.
I opened the same PR on my forked repository so that we can test the comment, you can see that here: https://github.com/sarahboyce/django/pull/1
Looks to work :tada: (@ngnpope I think that's what you wanted)

sarahboyce commented on .github/workflows/screenshots.yml:
```diff
@@ -0,0 +1,69 @@
+
+jobs:
+  selenium-sqlite-screenshots:
+    if: contains(github.event.pull_request.labels.*.name, 'selenium')
```
Yes I think we should. Can you create a label for me? 🙏

felixxm replied:
Added `screenshots 🖼️` with emoji :wink: .

sarahboyce replied:
Ahh amazing! Makes me very happy 🖼️❤️

sarahboyce commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -269,6 +271,10 @@ faster and more stable. Add the ``--headless`` option to enable this mode.
+For testing changes to the admin UI, the selenium tests can be run with the
+``--screenshots`` option enabled. Screenshots will be saved to the
+``tests/screenshots/`` directory.
+
```
Yeah I agree. Added an example in here 79648a4 👍

felixxm commented on django/test/selenium.py:
```diff
@@ -134,3 +137,82 @@ def disable_implicit_wait(self):
+
+                setattr(cls, test.__name__, test)
+
+    @contextmanager
+    def desktop(self):
+        window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(1280, 720)
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+
+    @contextmanager
+    def mobile(self):
+        window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(360, 800)
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+
+    @contextmanager
+    def rtl(self):
+        window_size = self.selenium.get_window_size()
+        get_language_bidi_mock = mock.patch(
+            "django.utils.translation.get_language_bidi", return_value=True
+        )
+        self.selenium.set_window_size(1280, 720)
+        get_language_bidi_mock.start()
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+            get_language_bidi_mock.stop()
```
I wonder if these 3 context managers could be moved to `SeleniumTestCase` and used as decorators directly on test methods (we have some tests that use `set_window_size()`), and reused here. This could standardize tests.

Also, we should use the same boundary widths as defined in admin CSS files, i.e.
- <= 767 (mobile)
- <= 1024 (small screens)
- standard

so maybe:

- `(767, 800)` -> `@mobile`
- `(1024, 768)` -> `@small_screen`
- `(1280, 720)` -> `@desktop`

We can do this in a separate PR.

felixxm replied:
So I would start with adding four context processors that can be used as test decorators, `@mobile`, `@small_screen`, `@desktop`, and `@rtl`.

sarahboyce replied:
Interesting suggestion. Originally I assumed these are only used/wanted when doing screenshots, if you agree with that then do you think the naming should be more exclicit around that it's related to taking a screenshot if this was updated? (like `@screenshot_mobile`)
If you think these are useful just in general, then I guess when doing the standard selenium tests we would want to re-run the test for each case regardless of whether it is taking a screenshot 🤔

felixxm replied:
> If you think these are useful just in general, then I guess when doing the standard selenium tests we would want to re-run the test for each case regardless of whether it is taking a screenshot 🤔

Yes, that's why I left this suggestion :smile: I imagine something like:
```python
class ChangeWindowSize:
    def __init__(self, width, height, selenium):
        self.new_size = (width, height)
        self.old_size = tuple(self.selenium.get_window_size().values())

    def __enter__(self):
        self.old_size = self.selenium.get_window_size()
        self.selenium.set_window_size(*self.new_size)
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.selenium.set_window_size(*self.old_size)

class SeleniumTestCase(...)

    @contextmanager
    def desktop(self):
        with ChangeWindowSize(1280, 720, self.selenium):
            yield

    @contextmanager
    def small_screen(self):
        with ChangeWindowSize(1024, 768, self.selenium):
            yield

    ...
```
and in tests, e.g.
```diff
diff --git a/tests/admin_widgets/tests.py b/tests/admin_widgets/tests.py
index 80bb6233ab..5ef69ce049 100644
--- a/tests/admin_widgets/tests.py
+++ b/tests/admin_widgets/tests.py
@@ -1076,6 +1076,7 @@ class DateTimePickerSeleniumTests(AdminWidgetSeleniumTestCase):
 
         self.assertEqual(len(selected), 0)
 
+    @self.small_screen
     def test_calendar_show_date_from_input(self):
         """
         The calendar shows the date from the input field for every locale
@@ -1083,7 +1084,6 @@ class DateTimePickerSeleniumTests(AdminWidgetSeleniumTestCase):
         """
         from selenium.webdriver.common.by import By
 
-        self.selenium.set_window_size(1024, 768)
         self.admin_login(username="super", password="secret", login_url="/")
 
         # Enter test data
```

The same context processors would be reused later in `_screenshot_cases`.

sarahboyce replied:
Ah thank you, I see what you're going for 👍

felixxm commented on django/test/selenium.py:
```diff
@@ -134,3 +137,82 @@ def disable_implicit_wait(self):
+        get_language_bidi_mock = mock.patch(
+            "django.utils.translation.get_language_bidi", return_value=True
+        )
+        self.selenium.set_window_size(1280, 720)
```
Can we reuse `desktop` here?

pauloxnet commented:
Do you think it's possible to take one image for each style mode (dark and light) for each screenshot so that we can test the admin UI in both style mode?

sarahboyce commented:
> Do you think it's possible to take one image for each style mode (dark and light) for each screenshot so that we can test the admin UI in both style modes?

Yes, everything is possible 😉 

Question though, do we need a "matrix" approach where every combination has a screenshot (so mobile vs desktop, ltr vs rtl, light vs dark - 2 x 2 x 2 = test is ran 8 times) or is it enough to test against a default which is what I am doing here (so rlt is only tested in desktop) which ends up with less test runs (including the dark mode, test would be ran 4 times).
Does that make sense?

felixxm commented:
Personally, I don't think it's worth doing.

pauloxnet commented:
> Personally, I don't think it's worth doing.

The risk is not noticing bugs introduced in the admin's dark theme mode.

felixxm commented:
> > Personally, I don't think it's worth doing.
> 
> The risk is not noticing bugs introduced in the admin's dark theme mode.

But those will be bugs in colors, not in a structure, am I right?

pauloxnet commented:
> > > Personally, I don't think it's worth doing.
> > 
> > The risk is not noticing bugs introduced in the admin's dark theme mode.
> 
> But those will be bugs in colors, not in a structure, am I right?

Not exactly, I would say that there would be bugs in the User eXperience and User Interface.

Potentially some sections of the interface could "disappear" simply by mistakenly setting the same color for a text and its background, or similar.

And this may not be visible by analyzing only the clear theme.

sarahboyce commented:
This might be my fault I answered @pauloxnet and then asked a seperate question.
@felixxm when you said you don't see the worth of doing it, did you just mean my 2 x 2 x 2 matrix question in which every combination should be tested or the dark mode screenshot or both?

felixxm commented:
> @felixxm when you said you don't see the worth of doing it, did you just mean my 2 x 2 x 2 matrix question in which every combination should be tested or the dark mode screenshot or both?

Yes, that's definitely not worth doing, maybe a separate `dark` marker to generate desktop screenshots in the dark mode.

sarahboyce commented:
Dark mode added 🕶️ (you can download the screenshots at the bottom here: https://github.com/django/django/actions/runs/6494619719/)

pauloxnet commented:
> Dark mode added 🕶️ (you can download the screenshots at the bottom here: https://github.com/django/django/actions/runs/6494619719/)

Great work @sarahboyce!

I see strange image height values (678px or 598px), different from what I read in your PR.

ngnpope commented on django/test/selenium.py:
```diff
@@ -118,6 +161,51 @@ def setUpClass(cls):
+    @contextmanager
+    def dark(self):
+        # Need to navigate to a page before executing a script.
+        self.selenium.get(self.live_server_url)
+        self.selenium.execute_script("localStorage.setItem('theme', 'dark');")
```
I forget with selenium... But do we need to wait for anything to load before executing the script here? Or does it work fine?

sarahboyce replied:
It appears to be working but maybe this is something I can do instead of navigating to a url 🤔

ngnpope commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -271,6 +271,32 @@ faster and more stable. Add the ``--headless`` option to enable this mode.
+
+To define when screenshots should be taken during a selenium test, the test
+class must use the ``django.test.selenium.screenshot_cases`` decorator with a
+list of supported screenshot types (currently ``"desktop"``, ``"mobile"``,
+``"rtl"`` and ``"dark"``). It can then call
```
You also have `small_screen` above? Or was the intent to get rid of that?

ngnpope replied:
Not sure whether it's worth stating that `rtl` and `dark` only operate with `desktop` size?

ngnpope commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -271,6 +271,32 @@ faster and more stable. Add the ``--headless`` option to enable this mode.
+            ...
+
+
+This generates three screenshots of the login page - one for a desktop screen,
+one for a mobile screen and one for right to left languages.
```
With the addition of `dark`, I think it's now **four**... 😂 

Maybe we can avoid specifying a number and just list the screenshots generated?

ngnpope reviewed (changes requested):
Hey @sarahboyce 👋🏻 

Finally got around to having a look. Nice to see this coming together.

ngnpope commented:
> I see strange image height values (678px or 598px), different from what I read in your PR.

@pauloxnet I guess this is because it is setting window size, not viewport size?

sarahboyce commented on .github/workflows/screenshots.yml:
```diff
@@ -0,0 +1,69 @@
+
+jobs:
+  selenium-sqlite-screenshots:
+    if: contains(join(github.event.pull_request.labels.*.name, '|'), 'screenshots')
```
It is because of the emoji (though I do love it) and I wasn't able to get it working by just putting the emoji into the string for some reason (I will commit this suggestion shortly to triple check)

ngnpope replied:
Might be worth checking how other projects that use emojis in labels handle this?

sarahboyce replied:
I think there might be a hidden character at the end of the label (if you hover over the different labels with emojis I think there's something at the end of the screenshot label: https://github.com/django/django/labels). 
@felixxm can you check the label and maybe delete the extra character? 🙏

felixxm replied:
> I think there might be a hidden character at the end of the label (if you hover over the different labels with emojis I think there's something at the end of the screenshot label: https://github.com/django/django/labels). @felixxm can you check the label and maybe delete the extra character? 🙏

There is nothing there (unfortunately).

sarahboyce commented on .github/workflows/screenshots.yml:
```diff
@@ -0,0 +1,69 @@
+            |-----------|-------------------------------------------------------------------------------------|
+            | Commit    | ${{ github.event.pull_request.head.sha }}                                           |
+            | Logs      | ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }} |
+          edit-mode: replace
```
Updated, you can see the example on my test PR: https://github.com/sarahboyce/django/pull/1

ngnpope replied:
Nearly there… But I think you forgot to add `id: create-screenshots` to the "Run Selenium tests with screenshots" step as the steps output stuff isn't working?

Would also be good to see how the extra list items get added to the comment when the commits are updated. We should see that if you update the branch to fix the above.

sarahboyce commented on django/test/selenium.py:
```diff
@@ -118,6 +161,51 @@ def setUpClass(cls):
+            try:
+                yield [desktop]
+            finally:
+                self.selenium.execute_script("localStorage.removeItem('theme');")
```
Any suggestions? I can add `_mode` to the end of each of them `dark_mode` `rtl_mode` etc

ngnpope replied:
Perhaps… 🤔

Could also go with `desktop_size` and `mobile_size`? (Although not sure `small_screen_size` would be better…)

Thoughts, @felixxm?

felixxm replied:
`_mode` and `_size` suffixes sound good to me.

felixxm replied:
Also, please move "size" helpers to a separate PR, they can be added and used without screenshots so I'd like to merge them first.

sarahboyce replied:
https://github.com/django/django/pull/17360 👍

felixxm replied:
Merged in f6629ee2c986d3bf59b4c1b3058f370a00bdc573. Please rebase.

felixxm commented:
@sarahboyce Thanks for all your efforts :+1: I would like to merge/discuss the remaining code in two steps:

1. new `screenshots` option to the `runtests.py`,
2. new GitHub action.

felixxm commented on django/test/selenium.py:
```diff
@@ -147,6 +174,36 @@ def mobile_size(self):
 
+    @contextmanager
+    def rtl(self):
+        get_language_bidi_mock = mock.patch(
+            "django.utils.translation.get_language_bidi", return_value=True
+        )
+        get_language_bidi_mock.start()
```
Can we use one of bidirectional languages instead of this mocking?
```python
with override_settings(LANGUAGE_CODE="he"):
    ...
```
or
```python
with override_settings(LANGUAGE_CODE=settings.LANGUAGES_BIDI[-1]):
    ...
```

sarahboyce replied:
This works 🎉 
It does mean that in tests when there is any selection based off link text `find_element(By.LINK_TEXT, "Example")` - it doesn't work out of the box 🤔 
What do you think? Is it better because the language is realistic? Or is it nice to keep it in english and have it go the other side via the mocking?

felixxm commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -271,6 +271,33 @@ faster and more stable. Add the ``--headless`` option to enable this mode.
+            self.take_screenshot("login")
+            ...
+
+
```
Remove an extra blank line.

felixxm commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -271,6 +271,33 @@ faster and more stable. Add the ``--headless`` option to enable this mode.
+
+This generates multiple screenshots of the login page - one for a desktop
+screen, one for a mobile screen, one for right to left languages on desktop
+and one for dark mode on desktop.
```
`versionchanged` annotation is missing, for example:

```
.. versionchanged:: 5.1

     The ``--screenshots`` option and `@screenshot_cases` decorator were added.
```

felixxm commented:
I moved the first commit with small edits (release notes, comments) to the #17374.

felixxm commented:
> I moved the first commit with small edits (release notes, comments) to the #17374.

Merged in be56c982c0805f62a2948d9d7a3e21215c352174.

felixxm commented:
@sarahboyce Many thanks! rebased.

felixxm commented:
I simplified the generated comment a bit to:

![image](https://github.com/django/django/assets/2865885/ab917ba7-86d2-4833-b54e-b706dbd47e1c)

ngnpope commented:
> I simplified the generated comment a bit to:
> 
> ![image](https://user-images.githubusercontent.com/2865885/276146335-ab917ba7-86d2-4833-b54e-b706dbd47e1c.png)

This reverts a bunch of what was discussed in [this comment](https://github.com/django/django/pull/16963#discussion_r1356743588), in particular reducing the visibility of screenshots taken on previous runs. Being able to easily access multiple downloads is useful for comparison. 

Maybe an alternative is not to edit existing comments, but just create new ones every time?

felixxm commented:
> Maybe an alternative is not to edit existing comments, but just create new ones every time?

Yes, I was thinking about the same :+1:

felixxm commented:
@ngnpope You can check the result in https://github.com/felixxm/django/pull/6.

felixxm commented:
This also make it easier to compare results when pushing fixes.

sarahboyce commented:
> > Maybe an alternative is not to edit existing comments, but just create new ones every time?
> 
> Yes, I was thinking about the same 👍

I'm not a fan of a new message with each commit. When you get enough comments, like this PR, conversations become hidden even when they're not resolved. I missed that Nick's suggestion earlier was to append to an existing message. Can try that? Or that a message is created in a thread? Worried a new message might pollute a review process 🤔

felixxm commented:
All good, but `date` is not there :thinking:

felixxm commented:
Finally, it works! :tada: https://github.com/felixxm/django/pull/9

sarahboyce commented:
> Finally, it works! 🎉 [felixxm#9](https://github.com/felixxm/django/pull/9)

Ah I love it!
